Split calculate-checksum parameter changeSetIdentifier into changeSetPath / changesetId / changeSetAuthor

JulienMa94 opened the pull request:
## Impact
- [ ] Bug fix (non-breaking change which fixes expected existing functionality)
- [ ] Enhancement/New feature (adds functionality without impacting existing logic)
- [X] Breaking change (fix or feature that would cause existing functionality to change)
 
`TypeBug`       
`APIBreakingChanges`
`newContributors` 


## Description
As described in #3814 the calculate-checksum command was not able to parse arguments which contains ``::`` characters. So the user is not able to use this command in conjuction with changeset Identifiers containing ``::``.
As suggested in the issue I have split the parameter changeSetIdentifier into three seperate parameters
- changeSetPath
- changeSetId
- changeSetAuthor

Herefor I made changes to the ``CalculateChecksumCommandStep`` and introduced the parameters mentioned above. I had to adjust the calculate-checksum API in Main and the Liquibase class.

So the user has to adjust his use of the calculate-checksum command with the new parameters like so.
``liquibase calculate-checksum --changeSetPath="changelog.xml"  --changeSetId="1::createTable" --changeSetAuthor="Me"``

Fixes #3814 

## Things to be aware of
- No Tests were written / adjusted until now
- New Contributor

## Things to worry about

## Additional Context

### Manual Test

**Run config**

![image](https://github.com/liquibase/liquibase/assets/37811304/492b336c-10a6-4a7e-854c-f5878935f668)

**Result**

![image](https://github.com/liquibase/liquibase/assets/37811304/811561d7-3de8-4c21-b65f-34f8971de67d)

MalloD12 commented on liquibase-standard/src/main/java/liquibase/Liquibase.java:
```diff
@@ -1227,26 +1227,17 @@ public void clearCheckSums() throws LiquibaseException {
      */
     @Deprecated
-    public final CheckSum calculateCheckSum(final String changeSetIdentifier) throws LiquibaseException {
+    public final CheckSum calculateCheckSum(final String changeSetPath, final String changeSetId, final String changeSetAuthor) throws LiquibaseException {
```
Even though this is a deprecated method it is a public API method and there might be people still using it, so I don't think it's a good idea to change its signature. I'd suggest to leave it as it was previously.

MalloD12 commented on liquibase-standard/src/main/java/liquibase/Liquibase.java:
```diff
@@ -1227,26 +1227,17 @@ public void clearCheckSums() throws LiquibaseException {
-     * @deprecated Use {link {@link CommandScope(String)}.
-     */
-    @Deprecated
-    public CheckSum calculateCheckSum(final String filename, final String id, final String author)
```
Instead, I would get back this one.

MalloD12 commented on liquibase-standard/src/main/java/liquibase/command/core/CalculateChecksumCommandStep.java:
```diff
@@ -20,22 +31,34 @@ public class CalculateChecksumCommandStep extends AbstractCommandStep {
+                                    .description("Changelog path in which the changeset is included")
+                                    .build();
+
+        CHANGESET_ID_ARG = builder.argument("changesetId", String.class)
```
Very minor thing, but let's rename this argument name to `changeSetId` to make it camel case and match with the two others.

MalloD12 commented on liquibase-standard/src/main/java/liquibase/command/core/CalculateChecksumCommandStep.java:
```diff
@@ -47,36 +70,36 @@ public List<Class<?>> requiredDependencies() {
 
-        List<String> parts = validateAndExtractParts(changeSetIdentifier, changeLogFile);
-        Scope.getCurrentScope().getLog(getClass()).info(String.format("Calculating checksum for changeset %s", changeSetIdentifier));
+        Scope.getCurrentScope().getLog(getClass()).info(String.format("Calculating checksum for changeset %s", changeSetId));
```
I think in this log message we should also include the other two arguments as well (`path` and `author`) since the three together compose the unique id of a changeset. Otherwise, there could be cases of having the same id multiple times in a changeset and this message can add a bit of confusion around.

MalloD12 commented on liquibase-standard/src/main/java/liquibase/command/core/CalculateChecksumCommandStep.java:
```diff
@@ -47,36 +70,36 @@ public List<Class<?>> requiredDependencies() {
+        ChangeSet changeSet = changeLog.getChangeSet(changeSetPath, changeSetAuthor, changeSetId);
         if (changeSet == null) {
-            throw new LiquibaseException(new IllegalArgumentException("No such changeSet: " + changeSetIdentifier));
+            throw new LiquibaseException(new IllegalArgumentException("No such changeSet: " + changeSetId));
```
Same here, I would also include the two other arguments as well.

MalloD12 commented on liquibase-standard/src/main/java/liquibase/Liquibase.java:
```diff
@@ -1227,26 +1227,17 @@ public void clearCheckSums() throws LiquibaseException {
      * @deprecated Use {link {@link CommandScope(String)}.
      */
     @Deprecated
-    public final CheckSum calculateCheckSum(final String changeSetIdentifier) throws LiquibaseException {
```
As this is an API breaking change, I think we need to return to its previous signature and also return back the one removed. I was thinking we can have something like the below code for backward compatibility and for new users just use CalculateCheckSumCommandStep way:  

```
/**
     * Calculate the checksum for a given identifier
     *
     * @deprecated Use {link {@link CommandScope(String)}.
     */
    @Deprecated
    public final CheckSum calculateCheckSum(final String changeSetIdentifier) throws LiquibaseException {
        String changeSetAttributes[] = changeSetIdentifier.split("::");
        //validate changeSet parameters and return an error or removed/ignore any other '::' occurrence when processing either a path, id or author. 
        return this.calculateCheckSum(changeSetAttributes[0], changeSetAttributes[1], changeSetAttributes[2]);
    }
    /**
     * Calculate the checksum for a given identifier
     *
     * @deprecated Use {link {@link CommandScope(String)}.
     */
    @Deprecated
    public CheckSum calculateCheckSum(final String changeSetPath, final String changeSetId, final String changeSetAuthor)
            throws LiquibaseException {
        CommandResults commandResults = new CommandScope("calculateChecksum")
                .addArgumentValue(DbUrlConnectionCommandStep.DATABASE_ARG, database)
                .addArgumentValue(CalculateChecksumCommandStep.CHANGESET_PATH_ARG, changeSetPath)
                .addArgumentValue(CalculateChecksumCommandStep.CHANGESET_ID_ARG, changeSetId)
                .addArgumentValue(CalculateChecksumCommandStep.CHANGESET_AUTHOR_ARG, changeSetAuthor)
                .addArgumentValue(CalculateChecksumCommandStep.CHANGELOG_FILE_ARG, this.changeLogFile)
                .execute();
        return commandResults.getResult(CalculateChecksumCommandStep.CHECKSUM_RESULT);
    }
```

JulienMa94 replied:
Reverted my changes and added your suggestion

MalloD12 commented on liquibase-standard/src/main/java/liquibase/command/core/CalculateChecksumCommandStep.java:
```diff
@@ -47,36 +70,36 @@ public List<Class<?>> requiredDependencies() {
 
-        List<String> parts = validateAndExtractParts(changeSetIdentifier, changeLogFile);
-        Scope.getCurrentScope().getLog(getClass()).info(String.format("Calculating checksum for changeset %s", changeSetIdentifier));
+        Scope.getCurrentScope().getLog(getClass()).info(String.format("Calculating checksum for changeset %s", changeSetId));
```
I think in this log message we should also include the other two arguments as well (path and author) since the three together compose the unique id of a changeset. Otherwise, there could be cases of having the same id multiple times in a changeset and this message can add a bit of confusion around.

JulienMa94 replied:
Added both arguments

MalloD12 commented on liquibase-standard/src/main/java/liquibase/command/core/CalculateChecksumCommandStep.java:
```diff
@@ -47,36 +70,36 @@ public List<Class<?>> requiredDependencies() {
+        ChangeSet changeSet = changeLog.getChangeSet(changeSetPath, changeSetAuthor, changeSetId);
         if (changeSet == null) {
-            throw new LiquibaseException(new IllegalArgumentException("No such changeSet: " + changeSetIdentifier));
+            throw new LiquibaseException(new IllegalArgumentException("No such changeSet: " + changeSetId));
```
Same here.

JulienMa94 replied:
Added both arguments here aswell

MalloD12 commented on liquibase-standard/src/main/java/liquibase/integration/commandline/Main.java:
```diff
@@ -1513,7 +1501,11 @@ protected void doMigration() throws Exception {
                 return;
             } else if (COMMANDS.CALCULATE_CHECKSUM.equalsIgnoreCase(command)) {
-                liquibase.calculateCheckSum(commandParams.iterator().next());
+                liquibase.calculateCheckSum(
```
I think here we should better perform the `calculateCheckSum` call by the command scope as below to avoid keep using the deprecated method and just leave the **Liquibase** class methods for those who use liquibase as a library: 

```
CommandResults commandResults = new CommandScope("calculateChecksum")
                .addArgumentValue(DbUrlConnectionCommandStep.DATABASE_ARG, database)
                .addArgumentValue(CalculateChecksumCommandStep.CHANGESET_PATH_ARG, changeSetPath)
                .addArgumentValue(CalculateChecksumCommandStep.CHANGESET_ID_ARG, changeSetId)
                .addArgumentValue(CalculateChecksumCommandStep.CHANGESET_AUTHOR_ARG, changeSetAuthor)
                .addArgumentValue(CalculateChecksumCommandStep.CHANGELOG_FILE_ARG, this.changeLogFile)
                .execute();
return commandResults.getResult(CalculateChecksumCommandStep.CHECKSUM_RESULT);
```

JulienMa94 replied:
Added the command scope call

MalloD12 commented on liquibase-cli/src/main/java/liquibase/integration/commandline/LiquibaseCommandLine.java:
```diff
@@ -104,7 +104,7 @@ private void cleanup() {
 
     public LiquibaseCommandLine() {
         this.legacyPositionalArguments = new HashMap<>();
-        this.legacyPositionalArguments.put("calculatechecksum", CalculateChecksumCommandStep.CHANGESET_IDENTIFIER_ARG.getName());
```
I think we should return this one to its previous state as well.

JulienMa94 replied:
Reverted to changeset identifier

MalloD12 commented:
Hi @JulienMa94,

Thank you for submitting this PR for us, I have left you a few comments. Please have a look at them and please let me know if you have any questions.

Thanks,
Daniel.

JulienMa94 commented:
@MalloD12 thx for your review. I will look into your suggested changes as soon I can find some freetime.
Greetings, Julien

JulienMa94 commented:
Hi @MalloD12,

I had some time to look into your suggestions and tried to refine the code to your suggestions. 

Unfortunately I have some problems with my local project setup and could not test my changes until now. 
At the moment I receive a NPE inside the Scope class at line 427 

https://github.com/liquibase/liquibase/blob/e885b06bb3bb87c1b165b6dd127b93bfdc315996/liquibase-standard/src/main/java/liquibase/Scope.java#L426-L431

I tried to debug this problem but I could not resolve it and I am a bit clueless if it is a problem related to my code changes or to my local run configuration.

Greetings
Julien

MalloD12 commented on liquibase-standard/src/main/java/liquibase/command/core/CalculateChecksumCommandStep.java:
```diff
@@ -20,22 +31,41 @@ public class CalculateChecksumCommandStep extends AbstractCommandStep {
-                .description("Changeset ID identifier of form filepath::id::author").build();
+                                    .description("The root changelog file").build();
+
+        CHANGESET_IDENTIFIER_ARG = builder.argument("changeSetIdentifier", String.class)
```
Hi @JulienMa94, I have been reviewing this PR yesterday and talked about it with the team and we have a few more things we might need to change. One of these is right now we are specifying the four arguments (`CHANGESET_IDENTIFIER_ARG`, `CHANGESET_PATH_ARG`, `CHANGESET_ID_ARG` and `CHANGESET_AUTHOR_ARG`) all defined as required. And I think this could be a bit ambiguous. What one of my team members suggest is we should make the four of them optional and then add a bit of logic in the `run()` method to control which of these arguments we will use. 

Doing the mentioned update will require us to update some of our command definition tests, plus in order to avoid confusion for other users I think we can add some examples like the ones below to command help. We can do this by calling `commandDefinition.setHelpFooter()` in the `adjustCommandDefinition()` method.

```
liquibase calculateCheckSum --changeSetIdentifier myFile::myId::myAuthor
```
 and 
```
liquibase calculatCheckSum --changesetId myId --changesetAuthor myAuthor --changesetPath myPath
```
If you need any help I'm here to help/assist you as you need, or I can even help doing these changes if you don't have any time constraints. 

Thanks,
Daniel.

MalloD12 commented on liquibase-standard/src/main/java/liquibase/command/core/CalculateChecksumCommandStep.java:
```diff
@@ -20,22 +31,41 @@ public class CalculateChecksumCommandStep extends AbstractCommandStep {
+
+        CHANGESET_IDENTIFIER_ARG = builder.argument("changeSetIdentifier", String.class)
+                                    .required()
+                                    .description("Changeset ID identifier of form filepath::id::author")
```
A minor thing. Can we use `ChangeSet` instead of `Changeset` argument descriptions? In this particular case, I would also remove the ID part as well.

MalloD12 commented:
> Hi @MalloD12,
> 
> I had some time to look into your suggestions and tried to refine the code to your suggestions.
> 
> Unfortunately I have some problems with my local project setup and could not test my changes until now. At the moment I receive a NPE inside the Scope class at line 427
> 
> https://github.com/liquibase/liquibase/blob/e885b06bb3bb87c1b165b6dd127b93bfdc315996/liquibase-standard/src/main/java/liquibase/Scope.java#L426-L431
> 
> I tried to debug this problem but I could not resolve it and I am a bit clueless if it is a problem related to my code changes or to my local run configuration.
> 
> Greetings Julien

I haven't faced this particular issue previously, but a team member told me he did and it got fixed for him by rebuilding the project. 

Thanks,
Daniel.

JulienMa94 commented:
Will try to rebuild it 👍🏽

JulienMa94 commented:
Hi @MalloD12,

I have implemented your suggestions and made the following parameters optional 

- changeSetIdentifier
- changeSetPath
- changeSetId
- changeSetAuthor

Inside the run method there is a case distinction if a single parameter (**changeSetIdentifier**) and multiple parameters (**changeSetPath, changeSetid, changeSetAuthor**) are passed by the user.
The changeSetIdentifier follows the orginial pattern `myFile::myId::myAuthor`.
I have added a help text to inform the user that there are two options for the Calculate Checksum command.


**Manual Tests:**

Single parameter config:
![Single_parameter_config](https://github.com/liquibase/liquibase/assets/37811304/6a7a6491-51d5-44c3-bc10-4ece70e68cff)

Single parameter result:
![Single_parameter_result](https://github.com/liquibase/liquibase/assets/37811304/a2f370e9-58fb-4fc4-b6de-e24d0f630b03)

Multiple parameters config:
![Multiple_parameter_config](https://github.com/liquibase/liquibase/assets/37811304/23248615-b514-488a-8a24-ee752cd0de82)

Multiple parameters result:
![Multiple_parameter_result](https://github.com/liquibase/liquibase/assets/37811304/cd7393ea-31a7-4a45-aeaf-e5a659a61ad5)

JulienMa94 commented:
> > Hi @MalloD12,
> > I had some time to look into your suggestions and tried to refine the code to your suggestions.
> > Unfortunately I have some problems with my local project setup and could not test my changes until now. At the moment I receive a NPE inside the Scope class at line 427
> > https://github.com/liquibase/liquibase/blob/e885b06bb3bb87c1b165b6dd127b93bfdc315996/liquibase-standard/src/main/java/liquibase/Scope.java#L426-L431
> > 
> > I tried to debug this problem but I could not resolve it and I am a bit clueless if it is a problem related to my code changes or to my local run configuration.
> > Greetings Julien
> 
> I haven't faced this particular issue previously, but a team member told me he did and it got fixed for him by rebuilding the project.
> 
> Thanks, Daniel.

Rebuilding did the trick, thx for the hint 👍🏼

MalloD12 commented:
Hi @JulienMa94,

Thank you for addressing the discussed changes. I reviewed them today and they look good to me. I have made a few minor text updates, plus fixed some command definition tests (as well as added one happy path test by making use of the three new arguments).  

We are having a build issue when trying to publish artifacts generated from a contributor branch and the team is already working on it. Before approving this PR I need to hold and wait until that issue is resolved, since once the build is successful some other tests should be triggered. I hope this gets fixed, so we can move forward with this one next week.

Again, thank you for all the effort you have been doing.

Thanks,
Daniel.

JulienMa94 commented:
> Hi @JulienMa94,
> 
> Thank you for addressing the discussed changes. I reviewed them today and they look good to me. I have made a few minor text updates, plus fixed some command definition tests (as well as added one happy path test by making use of the three new arguments).
> 
> We are having a build issue when trying to publish artifacts generated from a contributor branch and the team is already working on it. Before approving this PR I need to hold and wait until that issue is resolved, since once the build is successful some other tests should be triggered. I hope this gets fixed, so we can move forward with this one next week.
> 
> Again, thank you for all the effort you have been doing.
> 
> Thanks, Daniel.

Hi @MalloD12,

no problem I will be waiting and hoping that the issues will get fixed soon.
Thank you aswell for the constructive review and the work you did!

Have a nice week,
Julien

rberezen commented:
Thank you @JulienMa94 for the PR submission! I've taken a look and conducted tests with various scenarios involving different numbers of arguments—providing all, none, or only some of the required ones. In these cases, the error messages received were not particularly clear or user-friendly.

Encountered an unexpected error while executing Liquibase: The specified changeSet could not be identified. ChangeSet ID: null, Author: null, Path: null, .

Could I kindly request an enhancement to the error messaging for these scenarios? I'm thinking along the lines of:

Error encountered while parsing the command line. Required arguments missing: '--changeset-id', '--changeset-path'.

Error encountered while parsing the command line. '--changeset-identifier' cannot be provided alongside other changeset arguments: '--changeset-id', '--changeset-path', '--changeset-author'.

Could I seek some insights from @filipelautert and @MalloD12 on this matter? Your guidance would be greatly appreciated.

JulienMa94 commented:
Hi @rberezen,

thanks for the feedback. This enhancement sounds reasonable I will be able to work on the PR next week.

Greetings
Julien

JulienMa94 commented:
Hi @rberezen,

I added your suggestions for the error messages. So if the user provides all four or an ambigous amount e.g id, path and identifier of Arguments the following error message should be displayed:
`Error encountered while parsing the command line. '--changeset-identifier' cannot be provided alongside other changeset arguments: '--changeset-id', '--changeset-path', '--changeset-author'.`

If the user provides insufficent arguments e.g missing id & path the following message will be printed:
Error encountered while parsing the command line. Required arguments missing: '--changeset-id', '--changeset-path'.

Grettings
Julien

MalloD12 commented:
> Hi @rberezen,
> 
> I added your suggestions for the error messages. So if the user provides all four or an ambigous amount e.g id, path and identifier of Arguments the following error message should be displayed: `Error encountered while parsing the command line. '--changeset-identifier' cannot be provided alongside other changeset arguments: '--changeset-id', '--changeset-path', '--changeset-author'.`
> 
> If the user provides insufficent arguments e.g missing id & path the following message will be printed: `Error encountered while parsing the command line. Required arguments missing: '--changeset-id', '--changeset-path'.`
> 
> Grettings Julien

Thanks @JulienMa94 for implementing these changes! I think the below error message:
```
`Error encountered while parsing the command line. Required arguments missing: '--changeset-id', '--changeset-path'.`
```
is not totally accurate. Remember we change these arguments to not be required. I think it would be good to mention to users that if `--changeset-identifier` is provided then no other changeset arguments are needed, otherwise these three other arguments should be all present. 

Thanks,
Daniel.

JulienMa94 commented:
Hi @MalloD12 and @rberezen,

I refined the error message.
If the user provides insufficent arguments e.g missing id & path the following message will be printed:
```
Error encountered while parsing the command line. If --changeset-identifier is not provided than --changeset-id, --changeset-author and --changeset-path must be specified. Missing argument:  '--changeset-author', '--changeset-path'.
```

Greetings
Julien

MalloD12 reviewed (approved):
**Approved**.

Thank you, @JulienMa94 for taking care of this PR and addressing all changes we have been requesting you. I very appreciate your efforts.  

## Review and testing results:
Code changes look good to me. Happy path tests have been added. Which I think complementary with some extra manual test and maybe functional test should be good enough.

Build and most of the tests checks have been successfully executed, except for a MySQL job on the functional tests which will be addressed soon separately (for more details see #4863).

NOTE: PD [ticket](https://datical.atlassian.net/browse/PD-3700) (internal ticket) to reflect this changes on our docs has been created.  

### Things to be aware of:
- None

### Things to worry about:
- None

filipelautert commented on liquibase-standard/src/main/java/liquibase/command/core/CalculateChecksumCommandStep.java:
```diff
@@ -47,42 +77,102 @@ public List<Class<?>> requiredDependencies() {
+
+        final boolean isChangeSetIdentifierPassed = changeSetIdentifier != null;
+
+        validateIdentifierParameters(commandScope, changeSetIdentifier);
```
@JulienMa94 I extracted the parameters validation to a new method so it's easier to spot this change.
